src/csharp/Microsoft.Spark/Broadcast.cs

```diff
@@ -0,0 +1,243 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.IO;
+using System.Runtime.Serialization.Formatters.Binary;
+using Microsoft.Spark.Interop;
+using Microsoft.Spark.Interop.Ipc;
+
+namespace Microsoft.Spark
+{
+    /// <summary>
+    /// A broadcast variable. Broadcast variables allow the programmer to keep a read-only variable
+    /// cached on each machine rather than shipping a copy of it with tasks. They can be used, for
+    /// example, to give every node a copy of a large input dataset in an efficient manner. Spark
+    /// also attempts to distribute broadcast variables using efficient broadcast algorithms to
+    /// reduce communication cost.
+    /// </summary>
+    [Serializable]
+    public sealed class Broadcast<T> : IJvmObjectReferenceProvider
+    {
+        [NonSerialized]
+        private readonly JvmObjectReference _jvmObject;
+        private readonly long _bid;
+        [NonSerialized]
+        private readonly string _path;
+
+        internal Broadcast(SparkContext sc,
+            T value,
+            string localDir,
+            JvmObjectReference sparkContext)
+        {
+            var tempDir = Path.Combine(localDir, "sparkdotnet");
```

Comment:
Should this be set as a property of SparkContext ?  Similar to how pyspark is setting this up in the SparkContext constructor.

Reply:
I think Terry felt it was not needed.

Reply:
I think we can set this once in the sparkcontext, so we don't have to do this every time we create a broadcast var.

Reply:
I am -1 on polluting SparkContext for this and it's not a part of SparkContext on the Scala side either. Note that we don't always have to follow PySpark design/implementation if it doesn't make sense.

Reply:
If we want to refactor this, we can move it to some utils class (to mirror Scala side).

Reply:
Sorry, what would the utils class do in this context exactly?

Reply:
If you want to exactly mirror Scala side, this functionality should in some utils class. (check where `getLocalDir` is defined on the Scala side). But you don't need to do that in this PR.